kratos/processes/calculate_discontinuous_distance_to_skin_process.cpp

```diff
@@ -764,6 +809,141 @@ namespace Kratos
+
+            std::unordered_map<IndexType, IndexType> global_id_to_local;
+            for (IndexType i = 0; i < mNumNodes; ++i) {
+                global_id_to_local[pElem->GetGeometry()[i].Id()] = i;
+            }
+
+            auto elem_dist = pElem->GetValue(ELEMENTAL_DISTANCES);
+            std::vector<std::vector<IndexType>> cut_edges_vector;
+            const auto edges_container = pElem->GetGeometry().GenerateEdges();
+            for (std::size_t i_edge = 0; i_edge < mNumEdges; ++i_edge) {
+                IndexType edge_i_id = edges_container[i_edge][0].Id();
+                IndexType edge_j_id = edges_container[i_edge][1].Id();
+                double i_distance = elem_dist[global_id_to_local[edge_i_id]];
+                double j_distance = elem_dist[global_id_to_local[edge_j_id]];
+                bool is_edge_cut = (i_distance * j_distance) < 0.0;
+                std::vector<IndexType> edge_vector;
+                edge_vector.push_back(std::min(edge_i_id, edge_j_id));
+                edge_vector.push_back(std::max(edge_i_id, edge_j_id));
+                if (is_edge_cut) {
+                    cut_edges_vector.push_back(edge_vector);
+                }
+            }
+            return cut_edges_vector;
+        });
+
+        auto check_edge_against_distances = [&, get_neighbour_cut_edges] (const Element::GeometryType::GeometriesArrayType& rEdgeContainer, Vector& rCuteEdgeVector) {
+            for (std::size_t i_edge = 0; i_edge < mNumEdges; ++i_edge) {
+                // Edge initially cut with distances exactly 0
+                if (rCuteEdgeVector[i_edge] >= 0) {
+                    IndexType edge_i_id = std::min(rEdgeContainer[i_edge][0].Id(), rEdgeContainer[i_edge][1].Id());
+                    IndexType edge_j_id = std::max(rEdgeContainer[i_edge][0].Id(), rEdgeContainer[i_edge][1].Id());
+                    auto& neighbours_vector_0 = rEdgeContainer[i_edge][0].GetValue(NEIGHBOUR_ELEMENTS).GetContainer();
+                    auto& neighbours_vector_1 = rEdgeContainer[i_edge][1].GetValue(NEIGHBOUR_ELEMENTS).GetContainer();
+
+                    // Get neighbour elements in a single vector without repetitions.
+                    std::vector<Element::WeakPointer> all_neighbour_vector;
+                    all_neighbour_vector.reserve(neighbours_vector_0.size() + neighbours_vector_1.size());
+                    all_neighbour_vector.insert(all_neighbour_vector.end(), neighbours_vector_0.begin(), neighbours_vector_0.end());
+                    all_neighbour_vector.insert(all_neighbour_vector.end(), neighbours_vector_1.begin(), neighbours_vector_1.end());
+                    std::sort(all_neighbour_vector.begin(), all_neighbour_vector.end(), GlobalPointerCompare<Element>());
+                    auto new_end_it = std::unique(all_neighbour_vector.begin(), all_neighbour_vector.end(), GlobalPointerComparor<Element>());
+                    all_neighbour_vector.erase(new_end_it, all_neighbour_vector.end());
```

Comment:
If the order is not important, you can use `std::unordered_set<Element::WeakPointer> all_neighbour_vector;` and initialize it manually, that way you don't have to sort/unique/erase. Not 100% but I would say it will be faster. Something like:

```C++
std::unordered_set<Element::WeakPointer> all_neighbour_vector;

for(auto &neigh : rEdgeContainer[i_edge][0].GetValue(NEIGHBOUR_ELEMENTS).GetContainer()) {
    all_neighbour_vector.insert(neigh)
}

for(auto &neigh : rEdgeContainer[i_edge][1].GetValue(NEIGHBOUR_ELEMENTS).GetContainer()) {
    all_neighbour_vector.insert(neigh)
}

bool is_edge_cut = false;
for (auto& g_ptr_neigh : all_neighbour_vector) {
    if (!is_edge_cut) {
        auto cut_edges = get_neighbour_cut_edges.Get(g_ptr_neigh);
        for (auto cut_edge: cut_edges) {
            if (cut_edge[0]==edge_i_id && cut_edge[1]==edge_j_id) {
                is_edge_cut = true;
                break;
            }
        }
    }
}
```

Aside from this all ok :+1:

Reply:
True, done!

Had to add the Hasher and Comparator as well from your code:
` std::unordered_set<Element::WeakPointer, SharedPointerHasher<Element::WeakPointer>, SharedPointerComparator<Element::WeakPointer>> all_neighbour_set;`